dev/ProgressRing/ProgressRing.cpp

```diff
@@ -41,18 +47,98 @@ void ProgressRing::OnRangeBasePropertyChanged(const winrt::DependencyObject&, co
     UpdateSegment();
 }
 
+void ProgressRing::OnForegroundPropertyChanged(const winrt::DependencyObject&, const winrt::DependencyProperty&)
+{
+    ApplyLottieAnimation();
+}
+
+void ProgressRing::OnBackgroundPropertyChanged(const winrt::DependencyObject&, const winrt::DependencyProperty&)
+{
+    ApplyLottieAnimation();
+}
+
 void ProgressRing::OnStrokeThicknessPropertyChanged(const winrt::DependencyPropertyChangedEventArgs& args)
 {
+    ApplyLottieAnimation();
     UpdateRing();
 }
 
-winrt::Size ProgressRing::ComputeEllipseSize(double thickness, double actualWidth, double actualHeight)
+void ProgressRing::OnIsIndeterminatePropertyChanged(const winrt::DependencyPropertyChangedEventArgs& args)
+{
+    UpdateStates();
+}
+
+void ProgressRing::ApplyLottieAnimation()
+{
+    if (auto&& player = m_player.get())
+    {
+        // ProgressRing only accounts for ActualWidth to ensure that it is always a circle
+        const float diameter = static_cast<float>(ActualWidth());
+        const auto size = winrt::Size({ diameter, diameter });
+
+        const auto [foregroundColor, backgroundColor] = [
+            foreground = Foreground().try_as<winrt::SolidColorBrush>(),
+            background = Background().try_as<winrt::SolidColorBrush>()]()
+        {
+            winrt::Color foregroundColor;
+            winrt::Color backgroundColor;
+
+            if (foreground)
+            {
+                foregroundColor = foreground.Color();
+            }
+            else
+            {
+                foregroundColor = SharedHelpers::FindResource(s_DefaultForegroundThemeResourceName, winrt::Application::Current().Resources()).as<winrt::SolidColorBrush>().Color();
+            }
+
+            if (background)
+            {
+                backgroundColor = background.Color();
+            }
+            else
+            {
+                backgroundColor = SharedHelpers::FindResource(s_DefaultBackgroundThemeResourceName, winrt::Application::Current().Resources()).as<winrt::SolidColorBrush>().Color();
+            }
+
+            return std::make_tuple(foregroundColor, backgroundColor);
+        }();
+       
+        
+        player.Source(winrt::make<ProgressRingLoading>(StrokeThickness(), size, foregroundColor, backgroundColor));
+    }
+}
+
+void ProgressRing::UpdateStates()
+{
+    if (IsIndeterminate())
+    {
+        winrt::VisualStateManager::GoToState(*this, s_IndeterminateStateName, true);
+
+        if (auto&& player = m_player.get())
```

Comment:
>if (auto&& player = m_player.get()) [](start = 8, length = 35)

Can you avoid checking for the player by making it a precondition that UpdateStates is never called when there is no player?

Reply:
The go to state calls would still need to be called if we did not have a player.

---
In reply to: [370907912](https://github.com/microsoft/microsoft-ui-xaml/pull/1858#discussion_r370907912) [](ancestors = 370907912)